src/lib.rs

```diff
@@ -134,34 +139,245 @@ pub struct Price
+        // Uses at most 57 bits
+        let midprice = base_price * PD_SCALE / other_price;
+        let midprice_expo = PD_EXPO + base.expo - other.expo;
+
+        // Compute the confidence interval.
+        // This code uses the 1-norm instead of the 2-norm for computational reasons.
+        // The correct formula is midprice * sqrt(c_1^2 + c_2^2), where c_1 and c_2 are the
+        // confidence intervals in price-percentage terms of the base and other. This quantity
+        // is difficult to compute due to the sqrt, and overflow/underflow considerations.
+        // Instead, this code uses midprice * (c_1 + c_2).
+        // This quantity is at most a factor of sqrt(2) greater than the correct result, which
+        // shouldn't matter considering that confidence intervals are typically ~0.1% of the price.
+
+        // The exponent is PD_EXPO for both of these. Each of these uses 57 bits.
+        let base_confidence_pct: u64 = (base.conf * PD_SCALE) / base_price;
+        let other_confidence_pct: u64 = (other.conf * PD_SCALE) / other_price;
+
+        // at most 58 bits
+        let confidence_pct = base_confidence_pct + other_confidence_pct;
+        // at most 57 + 58 - 29 = 86 bits, with the same exponent as the midprice.
+        // FIXME: round this up. There's a div_ceil method but it's unstable (?)
+        let conf = ((confidence_pct as u128) * (midprice as u128)) / (PD_SCALE as u128);
+
+        // Note that this check only fails if an argument's confidence interval was >> its price,
+        // in which case None is a reasonable result, as we have essentially 0 information about the price.
+        if conf < (u64::MAX as u128) {
+          let m_i64 = midprice as i64;
+          // This should be guaranteed to succeed because midprice uses <= 57 bits
+          assert!(m_i64 >= 0);
+          Some(PriceConf {
+            price: m_i64,
+            conf: conf as u64,
+            expo: midprice_expo,
+          })
+        } else {
+          None
+        }
+      }
+      (_, _) => None
+    }
+  }
+
+  // FIXME Implement these functions
+  // The idea is that you should be able to get the price of a mixture of tokens (e.g., for LP tokens)
+  // using something like:
+  // price1.scale_to_exponent(result_expo).cmul(qty1, 0).add(
+  //   price2.scale_to_exponent(result_expo).cmul(qty2, 0)
+  // )
+  //
+  // Add two PriceConfs assuming the expos are ==
+  pub fn add(&self, other: PriceConf) -> Option<PriceConf> {
+    panic!()
+  }
+
+  // multiply by a constant
+  pub fn cmul(&self, c: u64, e: i32) -> Option<PriceConf> {
+    panic!()
+  }
+
+  /**
+   * Get a copy of this struct where the price and confidence
+   * have been normalized to be less than `MAX_PD_V_U64`.
+   * Returns `None` if `price == 0` before or after normalization.
+   * FIXME: tests
+   */
+  pub fn normalize(&self) -> Option<PriceConf> {
+    if self.price > 0 {
+      // BPF only supports unsigned division
```

Comment:
(I think this comment isn't actually true. We had some issue with this in the on-chain pyth program, but i just tested and it seems to work just fine here. will fix in the future)